AX: Move HeadingLevel and HierarchicalLevel to AXCoreObject

hoffmanjoshua opened the pull request:
#### bc34242b2e93638f8a33dfdb7e28a5dd97e96d31
<pre>
AX: Move HeadingLevel and HierarchicalLevel to AXCoreObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=292746">https://bugs.webkit.org/show_bug.cgi?id=292746</a>
<a href="https://rdar.apple.com/150967922">rdar://150967922</a>

Reviewed by Tyler Wilcock.

By caching aria-level, we are able to move the implementations of
headingLevel and hierarchicalLevel to a shared AXCoreObject
implementation. This also requires that the tags h1-h6 are cached
on the isolated tree.

As a follow up to this PR, I would like to share the way we use
tags on the live and isolated trees, to prevent methods like
`headingTagLevel` from also having separate implementations.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::headingLevel const):
(WebCore::AXCoreObject::hierarchicalLevel const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::headingTagLevel const):
(WebCore::AccessibilityNodeObject::headingLevel const): Deleted.
(WebCore::AccessibilityNodeObject::hierarchicalLevel const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::ariaLevel const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::headingTagLevel const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/294788@main">https://commits.webkit.org/294788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f91a0abdc99aade02a797b952448ea97fd484dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102847 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108013 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53488 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104886 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22840 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78205 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35167 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105853 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17687 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92815 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58539 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10850 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52845 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87341 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10917 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110388 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29984 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22090 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87189 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30348 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89010 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86806 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22134 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31646 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9362 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24246 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29911 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35233 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29719 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33046 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31281 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1069,6 +1069,45 @@ unsigned AXCoreObject::blockquoteLevel() const
+
+unsigned AXCoreObject::hierarchicalLevel() const
+{
+    int level = ariaLevel();
```
Typing this local as an `int` implies that `ariaLevel` could return a negative value (e.g. one could imagine a special-case `-1` value for missing `aria-level`). But from the header, it returns an `unsigned`. Any reason not to make this local an `unsigned` rather than an `int`?

hoffmanjoshua replied:
Yeah, this should be an `unsigned`—good catch!

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1069,6 +1069,45 @@ unsigned AXCoreObject::blockquoteLevel() const
+            return level;
+    }
+
+    unsigned tagLevel = headingTagLevel();
+    if (tagLevel > 0)
+        return tagLevel;
```
Another way to write this:

```cpp
if (unsigned tagLevel = headingTagLevel())
    return tagLevel;
```

But I'm fine leaving it as-is too, you can pick which one you prefer. Same applies for the `ariaLevel` checks in this function and the other, those could be written in this style too I think if that is our preference.

hoffmanjoshua replied:
I think I prefer it as-is, since it makes it a little more readable and explicit (IMO) that we don't want to handle `tagLevel == 0`.

hoffmanjoshua replied:
I think I prefer it as-is, since it makes it a little more readable and explicit (IMO) that we don't want to handle `tagLevel == 0`.

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1069,6 +1069,45 @@ unsigned AXCoreObject::blockquoteLevel() const
+    // We measure tree hierarchy by the number of groups that the item is within.
+    level = 1;
+    for (RefPtr parent = parentObject(); parent; parent = parent->parentObject()) {
+        AccessibilityRole parentRole = parent->roleValue();
```
This is a case where I think `auto` is OK. It's obvious from the function being called (`roleValue`) that this returns an `AccessibilityRole`, so spelling out the type isn't necessary.

hoffmanjoshua replied:
Updated this!

hoffmanjoshua replied:
Updated this!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -453,6 +463,9 @@ void AXIsolatedObject::initializeProperties(const Ref<AccessibilityObject>& axOb
         setProperty(AXProperty::CanBeMultilineTextField, canBeMultilineTextField(object));
     }
 
+    if (object.isHeading() || object.isTableRow() || object.isTreeItem())
```
This is a bit inefficient, as we already have `if (isTableRow())` and `if (isTreeItem())` blocks, so we're duplicating that work. Maybe we could add `bool isTableRow` and `bool isTreeItem` local variables and read them multiple times?

hoffmanjoshua replied:
Good idea, added those vars.

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1069,6 +1069,45 @@ unsigned AXCoreObject::blockquoteLevel() const
+    // Hierarchy leveling starts at 1, to match the aria-level spec.
+    // We measure tree hierarchy by the number of groups that the item is within.
+    level = 1;
+    for (RefPtr parent = parentObject(); parent; parent = parent->parentObject()) {
```
I know this was written this way before, not something you did. But `parent` isn't a great name for this local variable, since it's only our parent in the first iteration, and every one after that it's not. `ancestor` and `ancestorRole` would be more accurate.

hoffmanjoshua committed:
AX: Move HeadingLevel and HierarchicalLevel to AXCoreObject
https://bugs.webkit.org/show_bug.cgi?id=292746
rdar://150967922

Reviewed by Tyler Wilcock.

By caching aria-level, we are able to move the implementations of
headingLevel and hierarchicalLevel to a shared AXCoreObject
implementation. This also requires that the tags h1-h6 are cached
on the isolated tree.

As a follow up to this PR, I would like to share the way we use
tags on the live and isolated trees, to prevent methods like
`headingTagLevel` from also having separate implementations.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::headingLevel const):
(WebCore::AXCoreObject::hierarchicalLevel const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::headingTagLevel const):
(WebCore::AccessibilityNodeObject::headingLevel const): Deleted.
(WebCore::AccessibilityNodeObject::hierarchicalLevel const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::ariaLevel const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::headingTagLevel const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/294788@main
